src/headless/converse-muc.js

```diff
@@ -507,6 +507,42 @@ converse.plugins.add('converse-muc', {
                     }, null, 'message', 'groupchat', null, room_jid,
                     {'matchBareFromJid': true}
                 );
+                this.muc_notifications_handler = _converse.connection.addHandler(stanza => {
+                    const item = sizzle(`x[xmlns="${Strophe.NS.MUC_USER}"]`, stanza).pop();
+
+                    if (item) {
+                        const from = stanza.getAttribute("from");
+                        const type = stanza.getAttribute("type");
+                        const affiliation = item.getAttribute('affiliation');
+                        const jid = item.getAttribute('jid');
+
+                        const data = {
+                          'from': from,
+                          'nick': Strophe.getNodeFromJid(jid),
+                          'type': type,
+                          'states': [],
+                          'show': type == 'unavailable' ? 'offline' : 'online',
+                          'affiliation': affiliation,
+                          'role': item.getAttribute('role'),
+                          'jid': jid
+                        }
+
+                        const attributes = Object.assign(data, {
+                            'jid': Strophe.getBareJidFromJid(data.jid),
+                            'resource': Strophe.getResourceFromJid(data.jid)
+                        });
+                        const occupant = this.occupants.findOccupant({'jid': data.jid});
+
+                        if (occupant) {
+                          occupant.save(attributes);
```

Comment:
Looks like `attributes` must actually be `data`.

I think you should write a test for this change. That would have caught all the problems identified so far and it will also make sure that this feature doesn't accidentally get removed or broken in the future.
Here's a test that checks for affiliation changes via `presence` stanzas:

https://github.com/conversejs/converse.js/blob/f1e1e538b63eb72f755975d27912d99cf16670b7/spec/muc.js#L2841

Reply:
ok